galsim/phase_psf.py

```diff
@@ -755,15 +758,17 @@ def _prepareDraw(self):
             t, i = heappop(self._update_time_heap)
             self._seek(t)
             # Update that PSF
-            psf = self._pending[i]
-            psf._step()
-            # If that PSF's next possible update time doesn't extend past its exptime, then
-            # push it back on the heap.
-            t += psf.time_step
-            if t < psf.t0 + psf.exptime:
-                heappush(self._update_time_heap, (t, i))
-            else:
-                psf._finalize()
+            psfref = self._pending[i]
+            psf = psfref()
+            if psf is not None:
```

Comment:
I'm sure this is right, but why is this check required (both here and above)?  What does it mean for a weak_ref to return None?  Any why is it ok to not do anything in that case?

Reply:
Thanks Mike.  The `weakref` returns None if the referent no longer exists (was deleted, went out of scope, ...).  I was following the pattern described [here](https://docs.python.org/3/library/weakref.html#weak-reference-objects).

Reply:
I guess we must be getting some None's, since codecov thought you hit both branches.  Is that related to this commit: ed660d1?  So if some got deleted, but others are still around, then we want to skip them in this bit of code?

Reply:
Yep.  I added that commit b/c of codecov.

We do want to skip any deleted PSFs, since a) there's no supported API for the user to access them at this point (and the user has deleted their own handles to the objects), and b) the referents (`None`s) don't have the required `_step` or `_finalize` methods.

Looking at this piece of code again though, I realized two things: 1) There's no reason to advance the time of the phase screens for a PSF we don't draw.  So I moved that bit of code down into the if statement.  2) I don't really need both `_pending` and `_update_time_heap`, since the latter is just an index into the former.  So a rewrote this bit to use only the `_pending` attribute, which acts like the former `_update_time_heap` except that it stores the weakref to the PSF directly, instead of through an index into another list.